skbio/stats/composition.py

```diff
@@ -0,0 +1,384 @@
+Examples
+--------
+
+>>> import numpy as np
+
+Consider a very simple environment with only 3 species. The species
+in the environment are equally distributed and their proportions are
+equilvalent:
+
+>>> otus = np.array([1./3, 1./3., 1./3])
+
+Suppose that an antibiotic kills off half of the population for the
+first two species, but doesn't harm the third species. Then the
+perturbation vector would be as follows
+
+>>> antibiotic = np.array([1./2, 1./2, 1])
+
+And the resulting perturbation would be
+
+>>> perturb(otus, antibiotic)
+array([ 0.25,  0.25,  0.5 ])
+
+Reference
+---------
+.. [1] V. Pawlowsky-Glahn. "Lecture Notes on Compositional Data Analysis"
+.. [2] J. A. Martin-Fernandez. "Dealing With Zeros and Missing Values in
+       Compositional Data Sets Using Nonparametric Imputation"
+
+"""
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from __future__ import absolute_import, division, print_function
+import numpy as np
+import scipy.stats as ss
+
+
+def closure(mat):
+    """
+    Performs closure to ensure that all elements add up to 1.
+
+    Parameters
+    ----------
+    mat : array_like
+       a matrix of proportions where
+       rows = compositions
+       columns = components
+
+    Returns
+    -------
+    array_like, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    """
+    mat = np.atleast_2d(mat)
+    if np.any(mat < 0):
+        raise ValueError("Cannot have negative proportions")
+    if mat.ndim > 2:
+        raise ValueError("Input matrix can only have two dimensions or less")
+    mat = mat / mat.sum(axis=1, keepdims=True)
+    return mat.squeeze()
+
+
+def multiplicative_replacement(mat, delta=None):
+    """Replace all zeros with small non-zero values.
+
+    It uses the multiplicative replacement strategy, replacing zeros
+    with small positive values and performing a closure operation so
+    that the compositions still add up to 1
+
+    Parameters
+    ----------
+    mat: array_like
+       a matrix of proportions where
+       rows = compositions and
+       columns = components
+
+    Returns
+    -------
+    numpy.ndarray, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skbio.stats.composition import multiplicative_replacement
+    >>> X = np.array([[.2,.4,.4, 0],[0,.5,.5,0]])
+    >>> multiplicative_replacement(X)
+    array([[ 0.1875,  0.375 ,  0.375 ,  0.0625],
+           [ 0.0625,  0.4375,  0.4375,  0.0625]])
+
+    """
+    mat = np.atleast_2d(mat)
```

Comment:
Some of the checks here are the same as in `closure`. It'd be nice to move these checks into a private function that `closure` and `multiplicative_replacement` call to reduce duplication.